banzai_floyds/tests/test_background.py

```diff
@@ -0,0 +1,96 @@
+from banzai_floyds.background import fit_background, set_background_region, BackgroundFitter
+from banzai_floyds.extract import set_extraction_region
+from banzai_floyds.tests.utils import generate_fake_science_frame
+from banzai_floyds.utils.binning_utils import bin_data
+import numpy as np
+from banzai import context
+from collections import namedtuple
+from astropy.table import Table
+from numpy.polynomial.legendre import Legendre
+from banzai_floyds.utils.order_utils import get_order_2d_region
+
+
+def test_background_fitting():
+    np.random.seed(234515)
+    fake_frame = generate_fake_science_frame(include_sky=True)
+    binned_data = bin_data(fake_frame.data, fake_frame.uncertainty, fake_frame.wavelengths, fake_frame.orders)
+    fake_frame.binned_data = binned_data
+    fake_profile_width_funcs = [Legendre(fake_frame.input_profile_sigma,) for _ in fake_frame.input_profile_centers]
+    fake_frame.profile = fake_frame.input_profile_centers, fake_profile_width_funcs, None
+    fake_frame.background_windows = [[[-15, -5], [5, 15]], [[-15, -5], [5, 15]]]
+    set_background_region(fake_frame)
+    fake_frame.extraction_windows = [[-5.0, 5.0], [-5.0, 5.0]]
+    set_extraction_region(fake_frame)
+    fitted_background = fit_background(binned_data, background_order=3)
+    fake_frame.background = fitted_background
+    # If we are fitting to the noise, I think the residuals / uncertainty per pixel should
+    # follow a Gaussian distribution with sigma=1. So check cuts of the residual
+    # distribution rather than a single cutoff value in assert_allclose
+    # The residuals still look more correlated especially in the y-profile direction,
+    # but I guess that shouldn't be surprising given how we are fitting
+    in_order = fake_frame.orders.data > 0
+    residuals = fake_frame.background[in_order] - fake_frame.input_sky[in_order]
+    residuals /= fake_frame.uncertainty[in_order]
+    assert (np.abs(residuals) < 3).sum() > 0.99 * in_order.sum()
+    # We assert that only edge pixels can vary by 5 sigma due to edge effects
+    for order in [1, 2]:
+        order_region = get_order_2d_region(fake_frame.orders.data == order)
+        residuals = fake_frame.background[order_region][-2:2, -2:2] - fake_frame.input_sky[order_region][-2:2, -2:2]
```

Comment:
Previously, I was trying to require all pixels to have the residuals in the background fit be below some threshold. Given the large dynamic range in signal to noise that is not the correct metric. Instead, if we are fitting to the noise, the residuals/uncertainty should follow a unit gaussian distribution. So we now check that are only 1% of outliers larger than 3 sigma (similar to a gaussian) and no wild outliers < 5 sigma (skipping the edges which can have weird artifacts.)